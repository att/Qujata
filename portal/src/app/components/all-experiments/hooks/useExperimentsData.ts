import { FetchDataStatus, IHttp, useFetch } from '../../../shared/hooks/useFetch';
import { useEffect, useState } from 'react';
import { APIS } from '../../../apis';
import { useFetchSpinner } from '../../../shared/hooks/useFetchSpinner';
import { useErrorMessage } from '../../../hooks/useErrorMessage';
import { ITestRunResult, ITestRunResultData } from '../../../shared/models/test-run-result.interface';

export type TestRunSubset = Pick<ITestRunResultData, 'id' | 'algorithm' | 'iterations' | 'message_size'>;
export type Experiment = Pick<ITestRunResult, 'id' | 'name' | 'end_time'> & { test_runs: TestRunSubset[] };

export interface IUseExperimentsData {
  test_suites: Experiment[];
  status: FetchDataStatus;
}
export interface ExperimentData {
  id: number;
  name: string;
  algorithms: string[];
  iterations: number[];
  message_sizes: number[];
  end_time: number;
};

export function useExperimentsData(): IUseExperimentsData {
    const [allExperiments, setAllExperiments] = useState<Experiment[]>([]);
    const { get, data, cancelRequest, status, error }: IHttp<Experiment[]> = useFetch({ url: APIS.allExperiments });

    useFetchSpinner(status);
    useErrorMessage(error);
    useEffect(() => {
      get();  
      return cancelRequest;
    }, [get, cancelRequest]);


    useEffect(() => {
      if (data) {
        setAllExperiments(data);
      }
    }, [data, status, allExperiments]);

    return { test_suites: allExperiments, status };
}
