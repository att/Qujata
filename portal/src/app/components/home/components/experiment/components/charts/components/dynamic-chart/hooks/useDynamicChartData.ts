import { useEffect, useState } from "react";
import { AttSelectOption } from "../../../../../../../../../shared/components/att-select";
import { ITestRunResult } from "../../../../../../../../../shared/models/test-run-result.interface";

export interface IUseDynamicChartData {
  yAxiosOptions: AttSelectOption[];
}
export function useDynamicChartData(chartData: ITestRunResult): IUseDynamicChartData {
  const [yAxiosOptions, setYAxiosOptions] = useState<AttSelectOption[]>([]);

  useEffect(() => {
    const uniqueKeys = new Set<string>();

    for (const testRun of chartData.test_runs) {
      const results = testRun.results;
      for (const key in results) {
        uniqueKeys.add(key);
      }
    }

    if (uniqueKeys.size > 0) {
      setYAxiosOptions(Array.from(uniqueKeys).map(key => ({ label: convertLabelByCapitalLetter(key), value: key })));
    }
  }, [chartData]);

  return {
    yAxiosOptions,
  };
}

function convertLabelByCapitalLetter(str: string): string {
  return str
  .split('_') // split the string by underscore
  .map(word => word.charAt(0).toUpperCase() + word.slice(1)) // capitalize the first letter of each word
  .join(' '); // join the words back together with a space
}
