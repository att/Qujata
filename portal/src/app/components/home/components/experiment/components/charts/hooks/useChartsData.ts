import { useEffect, useState } from "react";
import { getKeysOfData, getLabels } from "../utils/chart.utils";
import { ITestRunResultData } from "../../../../../../../shared/models/test-run-result.interface";
import { ILineChartData } from "../models/line-chart-data.interface";
import { colors } from "../../../../../../dashboard/components/charts/LineChart/LineChart.const";
import { IExperimentData } from "../../../Experiment";

export interface IUseChartsData {
    barChartLabels: string[];
    barChartData: ITestRunResultData[];
    barChartKeysOfData: string[];
    lineChartData: ILineChartData;
}

function processedLineChartData(data: ITestRunResultData[], keysOfData: string[]): ILineChartData {
    const processedData = data.reduce((acc: {[key1: string]: {[key2: string]: number[]}}, curr: any) => {
        keysOfData.forEach((key: string) => {
          if (curr.results[key] !== undefined) {
            if (!acc[curr.algorithm]) {
              acc[curr.algorithm] = { [key]: [curr.results[key]] };
            } else {
              if (!acc[curr.algorithm][key]) {
                acc[curr.algorithm][key] = [curr.results[key]];
              } else {
                acc[curr.algorithm][key].push(curr.results[key]);
              }
            }
          }
        });
        return acc;
      }, {});

      const iterations: number[] = data.map((run: ITestRunResultData) => run.iterations);
      iterations.sort((a: number, b: number) => a - b);
      const uniqueIterations: Set<number> = new Set(iterations);

      const chartData = {
        labels:  Array.from(uniqueIterations),
        datasets: Object.keys(processedData).map((algorithm, index) => {
        const data = keysOfData.reduce((acc: {[key: string]: number[]}, key: string) => {
            acc[key] = processedData[algorithm][key];
            return acc;
        }, {});
            return {
                label: `${algorithm}       `,
                data: data,
                fill: false,
                backgroundColor: colors[index % colors.length],
                borderColor: colors[index % colors.length],
                borderWidth: 1,
            };
        })
      };
      
      return chartData;
}

export function useChartsData(props: IExperimentData): IUseChartsData {
    const [barChartLabels, setBarChartLabels] = useState<string[]>([]);
    const [barChartData, setBarChartData] = useState<ITestRunResultData[]>([]);
    const [barChartKeysOfData, setBarChartKeysOfData] = useState<string[]>([]);
    const [lineChartData, setLineChartData] = useState<ILineChartData>();
    
    useEffect(() => {
        if(props.data && props.data.test_runs.length > 0) {
            const testRuns: ITestRunResultData[] = props.data.test_runs;
            setBarChartData(testRuns);
            const labels: string[] = getLabels(testRuns);
            setBarChartLabels(labels);
            const keysOfData: string[] = getKeysOfData(testRuns[0].results);
            setBarChartKeysOfData(keysOfData);
            const lineChartData: ILineChartData = processedLineChartData(testRuns, keysOfData);
            setLineChartData(lineChartData);
        }
    }, [props.data]);

    return {
        barChartLabels,
        barChartData,
        barChartKeysOfData,
        lineChartData,
    } as IUseChartsData;
}
