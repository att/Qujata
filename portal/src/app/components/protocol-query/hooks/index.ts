export * from './useDuplicateData';
export * from './useGetAlgorithms';
export * from './useGetIterations';
export * from './useMessageSizeData';
