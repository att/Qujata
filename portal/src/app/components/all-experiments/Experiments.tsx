import { ReactNode, useCallback, useEffect, useMemo, useState } from 'react';
import styles from './Experiments.module.scss';
import cn from 'classnames';
import { ExperimentData, IUseExperimentsData, useExperimentsData } from './hooks';
import { FetchDataStatus, IHttp, useFetch } from '../../shared/hooks/useFetch';
import { Spinner, SpinnerSize } from '../../shared/components/att-spinner';
import { ALL_EXPERIMENTS_TABLE_EN } from './translate/en';
import { CellContext } from '@tanstack/react-table';
import { Table } from '../../shared/components/table';
import { Button, ButtonActionType, ButtonSize, ButtonStyleType } from '../../shared/components/att-button';
import { APIS } from '../../apis';
import { useNavigate } from 'react-router-dom';
import { useFetchSpinner } from '../../shared/hooks/useFetchSpinner';
import { useErrorMessage } from '../../hooks/useErrorMessage';
import { formatDistanceToNow } from 'date-fns';
import CheckedSvg from '../../../assets/images/checked.svg';
import UnCheckedSvg from '../../../assets/images/unchecked.svg';
import TrashSvg from '../../../assets/images/trash.svg';
import TrashHoverSvg from '../../../assets/images/trash-hover.svg';
import DuplicateSvg from '../../../assets/images/duplicate.svg';
import { DeleteExperimentModal } from '../home/components/experiment/components/delete-experiment-modal';
import { parseExperimentsData } from './utils/parse-experiments-data.utils';

const DeleteAriaLabel: string = ALL_EXPERIMENTS_TABLE_EN.BUTTONS.DELETE;
const DuplicateAriaLabel: string = ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.LINKS.DUPLICATE;

export const Experiments: React.FC = () => {
  const { test_suites, status }: IUseExperimentsData = useExperimentsData();
  const [openDeleteModal, setOpenDeleteModal] = useState<boolean>(false);
  const [checkedRows, setCheckedRows] = useState<Record<number, boolean>>({});
  const experimentsData = useMemo(() => (parseExperimentsData(test_suites) ?? []), [test_suites]);
  const navigate = useNavigate();

  const { post, status: deleteStatus, error: deleteError, cancelRequest: cancelRequestDelete }: IHttp<unknown>
    = useFetch<unknown>({ url: APIS.deleteExperiments });
  useFetchSpinner(deleteStatus);
  useErrorMessage(deleteError);
  useEffect(() => cancelRequestDelete, [cancelRequestDelete]);

  const handleDeleteClick: () => void = useCallback((): void => {
    setOpenDeleteModal(true);
  }, []);

  const handleCloseDeleteExperimentModal: (confirm?: boolean) => void = useCallback((confirm?: boolean): void => {
    if (confirm) {
      post({
        data: {
          ids: Object.keys(checkedRows).map((key: string) => parseInt(key))
        }
      });
    }
    setOpenDeleteModal(false);
  }, [post, checkedRows]);

  const handleCheckboxClick = useCallback((rowInfo: ExperimentData): void => {
    const rowId = rowInfo.id as number;
    setCheckedRows((prevState: Record<number, boolean>) => ({
      ...prevState,
      [rowId]: !prevState[rowId],
    }));
  }, []);

  const handleDuplicateClick = useCallback((row: ExperimentData) => {
    // Navigate to the Home Page
    navigate('/qujata', { state: { row } });
  }, [navigate]);

  const headers = useMemo(() => {
    const columnDefs = [
      { 
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.CHECKBOX,
        accessor: () => null,
        cell: (cellInfo: CellContext<ExperimentData, unknown>) => {
          const rowInfo: ExperimentData = cellInfo.row.original;
          return (
            <div className={styles.input_option} key={rowInfo.id}>
              <img
                data-testid={`${rowInfo.id}-checkbox-image`}
                className={styles.input_option_checkbox_icon}
                src={checkedRows[rowInfo.id!] ? CheckedSvg : UnCheckedSvg}
                alt='row-option'
                onClick={() => handleCheckboxClick(rowInfo)}
              />
              <input
                data-testid={`${rowInfo.id}-checkbox`}
                type='checkbox'
                id={rowInfo.id?.toString()}
                className={styles.input_form_item}
                onChange={() => handleCheckboxClick(rowInfo)}
              />
            </div>
          )
        }
      },
      {
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.EXPERIMENT_NAME.ID,
        name: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.EXPERIMENT_NAME.NAME,
        accessor: (row: ExperimentData) => row.name
      },
      {
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.ALGORITHMS.ID,
        name: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.ALGORITHMS.NAME,
        accessor: (row: ExperimentData) => row.algorithms?.join(', ')
      },
      {
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.ITERATIONS.ID,
        name: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.ITERATIONS.NAME, 
        accessor: (row: ExperimentData) => row.iterations?.join(', ')
      },
      { 
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.DATE.ID,
        name: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.DATE.NAME,
        accessor: (row: ExperimentData) => formatDistanceToNow(row.end_time as string, { addSuffix: true })
      },
      { 
        id: ALL_EXPERIMENTS_TABLE_EN.TABLE_COLUMNS.LINKS.DUPLICATE,
        accessor: () => null,
        cell: (cellInfo: CellContext<ExperimentData, unknown>) => (
          <Button
            ariaLabel={DuplicateAriaLabel}
            size={ButtonSize.NONE}
            styleType={ButtonStyleType.WRAPPER}
            actionType={ButtonActionType.BUTTON}
            onButtonClick={() => handleDuplicateClick(cellInfo.row.original)}
          >
            <img src={DuplicateSvg} alt={DuplicateAriaLabel} />
          </Button>
        )
      },
    ];
  
    return columnDefs.map(({ id, name, accessor, cell }) => ({
      id,
      header: () => <span>{name}</span>,
      accessor,
      cell: cell || ((cellInfo: CellContext<ExperimentData, unknown>) => <span>{cellInfo.getValue() as ReactNode}</span>)
    }));
  }, [checkedRows, handleCheckboxClick, handleDuplicateClick]);

  const checkedExperimentNames = experimentsData
    .filter((experiment: ExperimentData) => checkedRows[experiment.id])
    .map((experiment: ExperimentData) => experiment.name);

  return (
    <div className={styles.experiments_wrapper}>
      {status === FetchDataStatus.Fetching ? renderSpinner() : (
        <>
          <div className={styles.title_options_container}>
            <label className={styles.experiments_title}>{`${ALL_EXPERIMENTS_TABLE_EN.TITLE} (${test_suites.length})`}</label>
            {Object.values(checkedRows).some((value: boolean) => value) && (
              <Button
                ariaLabel={DeleteAriaLabel}
                size={ButtonSize.NONE}
                styleType={ButtonStyleType.WRAPPER}
                actionType={ButtonActionType.BUTTON}
                onButtonClick={handleDeleteClick}
              >
                <div className={styles.options_wrapper}>
                  <img className={cn(styles.trash_icon, styles.default_image)} src={TrashSvg} alt={DeleteAriaLabel} />
                  <img className={cn(styles.trash_icon, styles.hover_image)} src={TrashHoverSvg} alt={DeleteAriaLabel} />
                </div>
              </Button>
            )}
          </div>
          <Table className={styles.experiments_table} headers={headers} data={experimentsData} />
          {openDeleteModal && <DeleteExperimentModal name={checkedExperimentNames} onClose={handleCloseDeleteExperimentModal} />}
        </>
      )}
    </div>
  ); 
}

function renderSpinner() {
  return (
    <div className={styles.spinner_overlay}>
      <div className={styles.spinner_wrapper}>
      <Spinner size={SpinnerSize.MEDIUM} />
      </div>
    </div>
  );
}
