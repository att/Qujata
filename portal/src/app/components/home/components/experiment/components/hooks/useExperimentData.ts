import { useEffect, useState } from "react";
import { APIS } from "../../../../../../apis";
import { replaceParams } from "../../../../../../shared/utils/replaceParams";
import { useParams } from "react-router-dom";
import { ITestRunResult, ITestRunResultData } from "../../../../../../shared/models/test-run-result.interface";
import { TestRunUrlParams } from "../../../../../../shared/models/url-params.interface";
import { FetchDataStatus, IHttp, useFetch } from "../../../../../../shared/hooks/useFetch";
import { sortDataByAlgorithm } from "../charts/utils/test-run.utils";

export interface IUseExperimentData {
    data: ITestRunResultData[];
}

export function useExperimentData(): IUseExperimentData {
  const [testRunData, setTestRunData] = useState<ITestRunResultData[]>([]);
  const { testSuiteId } = useParams<TestRunUrlParams>();
  const testRunUrl: string = replaceParams(APIS.testRunResults, { testSuiteId });
  const { get, data, cancelRequest, status }: IHttp<ITestRunResult> = useFetch({ url: testRunUrl });
  
  useEffect(() => {
      get();
      return cancelRequest;
  }, [get, cancelRequest]);

  useEffect(() => {
      if (status === FetchDataStatus.Success && data) {
          const sortedData: ITestRunResultData[] = sortDataByAlgorithm(data.testRuns);
          setTestRunData(sortedData);
      }
  }, [status, data]);

  return {
      data: testRunData
  } as IUseExperimentData;
}
