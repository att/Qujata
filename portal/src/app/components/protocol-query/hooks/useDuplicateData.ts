import { useEffect } from 'react';
import { AttSelectOption } from '../../../shared/components/att-select';
import { ExperimentData } from '../../all-experiments/hooks';

export type DuplicateData = {
  data: ExperimentData | undefined,
  setDuplicateData: (data: any) => void,
  setExperimentName: (name: string) => void,
  setAlgorithms: (options: AttSelectOption[]) => void,
  setIterationsCount: (options: AttSelectOption[]) => void,
  setMessageSize: (options: AttSelectOption[]) => void
}

export const useDuplicateData = (duplicate: DuplicateData) => {
  useEffect(() => {
    if (duplicate.data) {
      const duplicateData = duplicate.data;
      if (duplicateData.name) {
        duplicate.setExperimentName(duplicateData.name);
      }
      if (duplicateData.algorithms) {
        const algorithmOptions = duplicateData.algorithms.map((algorithm: string) => {
          return { label: algorithm, value: algorithm } as AttSelectOption;
        });
        duplicate.setAlgorithms(algorithmOptions);
      }
      if (duplicateData.iterations) {
        const iterationsOptions = duplicateData.iterations.map((iteration: number) => {
          return { label: iteration.toString(), value: iteration.toString() } as AttSelectOption;
        });
        duplicate.setIterationsCount(iterationsOptions);
      }
      if (duplicateData.message_sizes) {
        const messageSizeOptions = duplicateData.message_sizes.map((messageSize: number) => {
          return { label: messageSize.toString(), value: messageSize.toString() } as AttSelectOption;
        });
        duplicate.setMessageSize(messageSizeOptions);
      }
      duplicate.setDuplicateData(undefined);
    }
  }, [duplicate]);
};
