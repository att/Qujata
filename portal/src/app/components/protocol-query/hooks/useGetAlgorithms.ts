import { AttSelectOption } from "../../../shared/components/att-select";
import { IHttp, useFetch } from "../../../shared/hooks/useFetch";
import { useEffect, useState } from "react";
import { APIS } from "../../../apis";

export interface IUseGetAlgorithms {
    options: AttSelectOption[];
}

interface IAlgorithm {
    classic: string[];
    hybrid: string[];
    quantumSafe: string[];
}
export function useGetAlgorithms(): IUseGetAlgorithms {
    const [options, setOptions] = useState<AttSelectOption[]>([]);
    const { get, data, cancelRequest }: IHttp<IAlgorithm> = useFetch({ url: APIS.algorithms });

    useEffect(() => {
        get();
        return cancelRequest;
    }, [get, cancelRequest]);


    useEffect(() => {
        if (data) {
            const classicOptions: AttSelectOption[] = data.classic.map((algo: string) => ({ label: algo, value: algo }));
            const hybridOptions: AttSelectOption[] = data.hybrid.map((algo: string) => ({ label: algo, value: algo }));
            const quantumSafeOptions: AttSelectOption[] = data.quantumSafe.map((algo: string) => ({ label: algo, value: algo }));
            const algorithmTitles: AttSelectOption[] = ['─────────── Classic ─────────────', '─────────── Hybrid ─────────────', '─────────── PQ ──────────────']
                .map((algo: string) => ({ label: algo, value: algo, isDisabled: true }));
            setOptions([
                algorithmTitles[0],
                ...classicOptions,
                algorithmTitles[1],
                ...hybridOptions,
                algorithmTitles[2],
                ...quantumSafeOptions
            ]);
        }
    }, [data]);

    return { options };
}
