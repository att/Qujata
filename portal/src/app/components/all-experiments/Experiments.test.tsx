import { render, fireEvent } from '@testing-library/react';
import { Experiments } from './Experiments';
import { useExperimentsData } from './hooks';
import { FetchDataStatus, useFetch } from '../../shared/hooks/useFetch';

jest.mock('./hooks');
jest.mock('../../shared/hooks/useFetch');
jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

describe('Experiments', () => {
  it('renders correctly', () => {
    (useExperimentsData as jest.Mock).mockReturnValue({
      test_suites: [{
        id: 15,
        name: "Experiment 1",
        end_time: 1705240065192,
        test_runs: [
          {
            id: 354,
            algorithm: "prime256v1",
            iterations: 100
          },
          {
            id: 355,
            algorithm: "prime256v1",
            iterations: 500
          }
        ]
      }],
      status: FetchDataStatus.Fetching,
    });
    (useFetch as jest.Mock).mockReturnValue({
      post: jest.fn(),
      status: FetchDataStatus.Fetching,
      error: null,
      cancelRequest: jest.fn(),
    });

    const { container } = render(<Experiments />);
    expect(container).toMatchSnapshot();
  });
});