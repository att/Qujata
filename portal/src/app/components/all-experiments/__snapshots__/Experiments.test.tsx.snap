// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experiments renders correctly 1`] = `
<div>
  <div
    class="experiments_wrapper"
  />
</div>
`;
